File: src/main/java/net/snipsniper/systray/PopupMenuButton.java
package net.snipsniper.systray;

import net.snipsniper.utils.IFunction;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.image.BufferedImage;
import java.util.ArrayList;

public class PopupMenuButton extends JMenuItem {
    private IFunction onClick;
    private boolean isMenuChild = false;

    public PopupMenuButton(String title, BufferedImage icon, JFrame popup, IFunction function, ArrayList<PopupMenu> closeWhenClicked) {
        setText(title);
        setIcon(getPopupIcon(icon));
        onClick = function;
        addActionListener(e -> {
            popup.setVisible(false);
            if(closeWhenClicked != null)
                for(PopupMenu menu : closeWhenClicked)
                    menu.setPopupMenuVisible(false);
            if(onClick != null)
                onClick.run();
        });
        addMouseListener(new MouseAdapter() {
            @Override
            public void mouseEntered(MouseEvent e) {
                super.mouseEntered(e);
                setArmed(true);
                if(!isMenuChild)
                    for(PopupMenu menu : closeWhenClicked)
                        menu.setPopupMenuVisible(false);
            }

            @Override
            public void mouseExited(MouseEvent e) {
                super.mouseExited(e);
                setArmed(false);
            }
        });
    }

    public void setFunction(IFunction function) {
        onClick = function;
    }

    public void setIsMenuChild(boolean value) {
        isMenuChild = value;
    }

    public ImageIcon getPopupIcon(BufferedImage image) {
        return new ImageIcon(image.getScaledInstance(16,16, Image.SCALE_DEFAULT));
    }

}

File: src/main/kotlin/net/snipsniper/systray/PopupMenuButton.kt
package net.snipsniper.systray

import net.snipsniper.utils.getImage
import net.snipsniper.utils.scaled
import net.snipsniper.utils.toImageIcon
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.image.BufferedImage
import javax.swing.ImageIcon
import javax.swing.JDialog
import javax.swing.JMenuItem

class PopupMenuButton(title: String, icon: BufferedImage, popup: JDialog, private var onClick: (() -> (Unit))?, closeWhenClicked: ArrayList<PopupMenu>?): JMenuItem() {
    private var isMenuChild = false

    constructor(title: String, iconPath: String, popup: JDialog, onClick: (() -> (Unit))?, closeWhenClicked: ArrayList<PopupMenu>?) : this(title, iconPath.getImage(), popup, onClick, closeWhenClicked)

    init {
        text = title
        this.icon = getPopupIcon(icon)
        addActionListener {
            popup.isVisible = true
            closeWhenClicked?.forEach { it.isPopupMenuVisible = false }
            onClick?.invoke()
        }
        addMouseListener(object: MouseAdapter(){
            override fun mouseEntered(e: MouseEvent?) {
                isArmed = true
                if(!isMenuChild) closeWhenClicked?.forEach { it.isPopupMenuVisible = false }
            }
            override fun mouseExited(e: MouseEvent?) = kotlin.run { isArmed = false }
        })
    }

    fun setIsMenuChild(value: Boolean) = kotlin.run { isMenuChild = value }
    private fun getPopupIcon(image: BufferedImage): ImageIcon = image.scaled(16, 16).toImageIcon()
}